genie-core/src/main/java/com/netflix/genie/core/services/impl/CacheGenieFileTransferService.java

```diff
@@ -0,0 +1,152 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package com.netflix.genie.core.services.impl;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.netflix.genie.common.exceptions.GenieException;
+import com.netflix.genie.common.exceptions.GenieServerException;
+import com.netflix.genie.core.services.FileTransfer;
+import com.netflix.genie.core.services.FileTransferFactory;
+import com.netflix.spectator.api.Registry;
+import lombok.extern.slf4j.Slf4j;
+import org.hibernate.validator.constraints.NotBlank;
+
+import javax.validation.constraints.NotNull;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.UUID;
+import java.util.function.ToDoubleFunction;
+
+/**
+ * Caches the downloaded file from the remote location.
+ * Created by amajumdar on 7/22/16.
+ */
+@Slf4j
+public class CacheGenieFileTransferService extends GenieFileTransferService {
+    //File cache location
+    private final String baseCacheLocation;
+    //File transfer service to get/put files on a local system
+    private final FileTransfer localFileTransfer;
+    //File cache
+    private final LoadingCache<String, File> fileCache = CacheBuilder.newBuilder()
+            .recordStats()
+            .build(
+                    new CacheLoader<String, File>() {
+                        public File load(
+                                @NotNull
+                                final String path) throws GenieException {
+                            return loadFile(path);
+                        }
+                    });
+
+    /**
+     * Constructor.
+     *
+     * @param fileTransferFactory file transfer implementation factory
+     * @param baseCacheLocation file cache location
+     * @param localFileTransfer Local file transfer service
+     * @param registry spectator registry
+     * @throws GenieException If there is any problem
+     */
+    public CacheGenieFileTransferService(
+            @NotNull final FileTransferFactory fileTransferFactory,
+            @NotNull final String baseCacheLocation,
+            @NotNull final FileTransfer localFileTransfer,
+            @NotNull final Registry registry) throws GenieException {
+        super(fileTransferFactory);
+        this.baseCacheLocation = baseCacheLocation;
+        createDirectories(baseCacheLocation);
+        this.localFileTransfer = localFileTransfer;
+        registry.gauge("genie.jobs.file.cache.hitRate", fileCache,
+                (ToDoubleFunction<LoadingCache<String, File>>) value -> value.stats().hitRate());
+        registry.gauge("genie.jobs.file.cache.missRate", fileCache,
```

Comment:
Since these cache's aren't used by only job files (in fact we don't use them for any job files really just clusters, commands and apps right?) I would rename metric to more closely match the ones I did for s3: genie.files.s3.download.timer... just under ```genie.files.cache.whatever```